lib/openhab/dsl/items/group_item.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require 'forwardable'
+require 'java'
+require 'openhab/dsl/items/item_command'
+require 'openhab/dsl/items/item_delegate'
+require 'openhab/core/entity_lookup'
+
+module OpenHAB
+  module DSL
+    module Items
+      #
+      # Delegator to OpenHAB Group Item
+      #
+      class GroupItem
+        extend OpenHAB::DSL::Items::ItemCommand
+        extend OpenHAB::DSL::Items::ItemDelegate
+        include Enumerable
+
+        def_item_delegator :@group_item
+
+        #
+        # Create a new GroupItem
+        #
+        # @param [Java::Org::openhab::core::items::GroupItem] group_item OpenHAB GroupItem to delegate to
+        #
+        def initialize(group_item)
+          @group_item = group_item
+
+          item_missing_delegate { @group_item }
+          add_state_methods
+          add_command_methods
+        end
+
+        #
+        # Iterates through the direct members of the Group
+        #
+        def each(&block)
+          OpenHAB::Core::EntityLookup.decorate_items(@group_item.members.to_a).each(&block)
+        end
+
+        #
+        # Get all members of the group recursively
+        #
+        # @return [Array] An Array containing all descendants of the Group
+        def all_members
+          OpenHAB::Core::EntityLookup.decorate_items(@group_item.all_members.to_a)
+        end
+
+        private
+
+        #
+        # Add state methods for the Groups base item type
+        #
+        def add_state_methods
+          data_enums = get_accepted_data_types.select(&:is_enum)
```

Comment:
Do we want to support that face that group type can change distinct from the rule runtime?

Reply:
Not sure I follow. Isn't the object recreated on every call to const_missing? Or is the result of those calls saved in memory when the rule is parsed and reused on every rule execution? In the latter case this would probably need to be reworked somehow.

Reply:
Made a test rule (where Switches is a group item):
```ruby
rule 'test' do
  every 5.seconds
  run do
    logger.info(Switches.__id__)
    logger.info(Switches.__id__)
  end
end
```
Which confirms that the object gets recreated on every call, so even if the underlying item should change between calls, the object gets re-instantiated and the correct methods should get created.

Reply:
Added a test to verify that this works as intended (if a group changes type that is reflected even in running rules)

Reply:
Ah, now I understand, good point. But wouldn't this be a problem with regular items as well? If it's saved to an array as a SwitchItem and then changed to a DimmerItem the object in the array would still be a SwitchItem, right?

Either case since it delegates to the base item (which is decorated on per-call basis) it should work even with those methods are removed.

But OH also lets you e.g. send a ON command to a group of Switches and Dimmers even if the group hasn't got a type, which wouldn't work with the`on` method.

If I would add `method_missing` would this shadow the one defined in ItemDelegate? Which one would get called first?

Reply:
Seems like the state methods aren't reliable for type-less groups anyway, since the group doesn't update its state if no aggregation function is defined, will remove them so we don't confuse users.